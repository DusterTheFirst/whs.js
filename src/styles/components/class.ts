/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "@whsha/classes/v2/block";
import styled from "styled-components/native";
import { getDisplayColorForBlock } from "../blockColor";
import { DimText } from "./common";

/** Text to be placed on the left of the class display */
export const LeftClassText = styled(DimText)({
    flex: 1,
    textAlign: "left",
});

/** Text to be placed on the right of the class display */
export const RightClassText = styled(DimText)({
    flex: 1,
    textAlign: "right",
});

/** Text to be placed for display as times */
export const TimesText = styled(DimText)({
    paddingLeft: 5,
    textAlign: "right"
});

/** Properties for a styled component with a class color prop */
interface IClassColoredComponentProps {
    /** The class color to show */
    classColor?: BlockColor;
}

/** The text to show the colorblind labels with */
export const ColorblindLabelText = styled.Text<IClassColoredComponentProps>(
    (props) => ({
        color:
            getDisplayColorForBlock(props.classColor)
            ?? props.theme.colors.dimText,
        fontWeight: "bold",
        paddingLeft: 10,
        paddingRight: 10,
        textAlign: "center"
    })
);

/** The text to show the title of a class */
export const ClassTitleText = styled(LeftClassText)<IClassColoredComponentProps>(
    (props) => ({
        color:
            getDisplayColorForBlock(props.classColor)
            ?? props.theme.colors.title,
        fontSize: 25,
        fontWeight: "bold",
    })
);

/** A view for placing an item in the center of a class view */
export const MiddleClassView = styled.View({
    alignItems: "center",
    flex: 1,
    justifyContent: "center",
    maxWidth: "75px"
});

/** A class view row */
export const ClassViewRow = styled.View({
    flexDirection: "row",
    flexWrap: "nowrap",
    justifyContent: "space-between"
});

/** The info row for the class view */
export const ClassViewInfoRow = styled(ClassViewRow)({
    marginTop: 5
});

/** Props for a styled component that can have a dynamic flex state */
interface IFlexableProps {
    /** The flex state */
    doFlex: boolean;
}

/** The view to surround the times part of the class */
export const TimesView = styled.View<IFlexableProps>(props => ({
    alignItems: "center",
    flex: props.doFlex ? 1 : undefined,
    flexDirection: "row",
    justifyContent: "flex-end"
}));

/** The view for containing a class */
export const ClassContainerView = styled.View(props => ({
    alignItems: "center",
    backgroundColor: props.theme.colors.classBackground,
    justifyContent: "center",
    padding: 10,
    width: "100%"
}));