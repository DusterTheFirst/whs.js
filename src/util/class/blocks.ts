/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "../blocks/blockColor";
import { IAdvisedClass, IIrregular } from "./primitives";
import { ClassType, IClassType } from "./type";

/** The most basic information needed to identify a major */
export interface IMajor extends IAdvisedClass, IClassType<ClassType.Major> {
    /** The name of the major */
    name: string;
    /** The block color which the major meets */
    block: BlockColor;
    /** If the major has a Lab block or not */
    lab: boolean;
}

/** The most basic information needed to identify an elective */
export interface IElective extends IAdvisedClass, IIrregular, IClassType<ClassType.Elective> {
    /** The name of the elective */
    name: string;
    /** The block color which the elective meets (None = 730) */
    block: BlockColor;
}

/** The most basic information needed to identify a DR */
export interface IDR extends IAdvisedClass, IIrregular, IClassType<ClassType.DR> {
    /** The block color which the elective meets */
    block: BlockColor;
}

/** The most basic information needed to identify a Free block */
export interface IFree extends IIrregular, IClassType<ClassType.Free> {
    /** The block color which the elective meets */
    block: BlockColor;
}