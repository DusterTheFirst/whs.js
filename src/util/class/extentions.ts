/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { Dayjs } from "dayjs";

/**
 * Metadata that is needed to understand where the class will
 * be placed in the schedule for each day but not needed to display the class
 *
 * (Internal Use)
 */
export interface IClassMeta {
    /** The class UUID for identifying the class */
    uuid: Readonly<string>;
}

/** Information that is displayed to the user but never stored locally */
export interface ITimes {
    /** The start time */
    start: Dayjs;
    /** The end time */
    end: Dayjs;
}