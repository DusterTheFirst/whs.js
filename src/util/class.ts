/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { Dayjs } from "dayjs";
import { BlockColor } from "./blocks/blockColor";

export default interface IClass {
    teacher: string;
    room: number;
    name: string;
    block: BlockColor;
    meets: number;
}

export interface ITimes {
    start: Dayjs;
    end: Dayjs;
}