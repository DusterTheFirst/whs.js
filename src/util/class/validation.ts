/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "../blocks/blockColor";
import { SchoolDay } from "../calendar/types";
import { ValidationError, ValidationWarning } from "../hooks/classes/useClasses";
import ProblemMap from "../problemMap";
import { IDR, IMajor, IMinor } from "./full";
import { irregularMeetCount, IrregularMeetDays, irregularMeetDays, irregularMeetJoin } from "./primitives";

export function validateMajors(problems: ProblemMap<string, ValidationError, ValidationWarning>, majors: IterableIterator<IMajor>): Set<BlockColor> {
    // Store existing color blocks
    const majorColors = new Set<BlockColor>();

    // Loop through all majors
    for (const major of majors) {
        // Check if one with the same major already exists.
        if (majorColors.has(major.block)) {
            // If it does, return an error
            problems.addError(major.uuid, ValidationError.MajorHasDuplicateBlockColor);
        } else {
            // If not, add it to the list and keep on going
            majorColors.add(major.block);
        }

        // Make sure a block has been specified
        if (major.block === BlockColor.None) {
            problems.addError(major.uuid, ValidationError.MajorMissingBlockColor);
        }

        // Make sure a name has been specified
        if (major.name.length === 0) {
            problems.addWarn(major.uuid, ValidationWarning.MissingName);
        }

        // Make sure a room has been specified
        if (major.room.length === 0) {
            problems.addWarn(major.uuid, ValidationWarning.MissingRoom);
        }

        // Make sure a teacher has been specified
        if (major.teacher.length === 0) {
            problems.addWarn(major.uuid, ValidationWarning.MissingTeacher);
        }
    }

    return majorColors;
}

export function validateMinors(problems: ProblemMap<string, ValidationError, ValidationWarning>, majorColors: Set<BlockColor>, minors: IterableIterator<IMinor>): Map<BlockColor, IrregularMeetDays> {
    // Store existing minor blocks
    const minorDays = new Map<BlockColor, IrregularMeetDays>();

    // Loop through all minors
    for (const minor of minors) {
        // Make sure the class meets more than once
        if (irregularMeetCount(minor) === 0) {
            problems.addError(minor.uuid, ValidationError.MissingMeetDay);
        }
        // Make sure the class does not meet all blocks for a color (Not Irregular)
        else if (irregularMeetCount(minor) >= 5 && minor.block !== BlockColor.None) {
            problems.addError(minor.uuid, ValidationError.MeetsEveryDay);
        }

        // Make sure the minor does not conflict with a major
        if (majorColors.has(minor.block)) {
            problems.addError(minor.uuid, ValidationError.MinorConflictWithMajor);
        }

        // Get the days that minors with the same color block meet
        let meetDays = minorDays.get(minor.block);
        // If no days exist, set it to a default
        if (meetDays === undefined) {
            meetDays = {
                [SchoolDay.One]: false,
                [SchoolDay.Two]: false,
                [SchoolDay.Three]: false,
                [SchoolDay.Four]: false,
                [SchoolDay.Five]: false,
                [SchoolDay.Six]: false,
                [SchoolDay.Seven]: false,
            };
        }

        // Make sure the minor does not overlap any existing days for the color
        if (irregularMeetDays({ meets: meetDays }).some(x => irregularMeetDays(minor).includes(x))) {
            problems.addError(minor.uuid, ValidationError.MinorConflictWithMinor);
        }

        // Save the meeting days by joining them with the others for the same color
        minorDays.set(minor.block, irregularMeetJoin(meetDays, minor.meets));

        // Make sure a name has been specified
        if (minor.name.length === 0) {
            problems.addWarn(minor.uuid, ValidationWarning.MissingName);
        }

        // Make sure a room has been specified
        if (minor.room.length === 0) {
            problems.addWarn(minor.uuid, ValidationWarning.MissingRoom);
        }

        // Make sure a teacher has been specified
        if (minor.teacher.length === 0) {
            problems.addWarn(minor.uuid, ValidationWarning.MissingTeacher);
        }
    }

    return minorDays;
}

export function validateDRs(problems: ProblemMap<string, ValidationError, ValidationWarning>, majorColors: Set<BlockColor>, minorDays: Map<BlockColor, IrregularMeetDays>, drs: IterableIterator<IDR>): Map<BlockColor, IrregularMeetDays> {
    // Store existing DR blocks
    const drDays = new Map<BlockColor, IrregularMeetDays>();

    // Loop through all minors
    for (const dr of drs) {
        // Make sure the class meets more than once
        if (irregularMeetCount(dr) === 0) {
            problems.addError(dr.uuid, ValidationError.MissingMeetDay);
        }
        // Make sure the class does not meet all blocks for a color (Not Irregular)
        else if (irregularMeetCount(dr) >= 5 && dr.block !== BlockColor.None) {
            problems.addError(dr.uuid, ValidationError.MeetsEveryDay);
        }

        // Make sure the minor does not conflict with a major
        if (majorColors.has(dr.block)) {
            problems.addError(dr.uuid, ValidationError.DRConflictWithMajor);
        }

        // Get the days that the minors with the same color block meet
        const minorMeetDays = minorDays.get(dr.block);

        // Make sure the dr does not overlap any existing minors for the color
        if (minorMeetDays !== undefined && irregularMeetDays({ meets: minorMeetDays }).some(x => irregularMeetDays(dr).includes(x))) {
            problems.addError(dr.uuid, ValidationError.DRConflictWithMinor);
        }

        // Get the days that drs with the same color block meet
        let meetDays = drDays.get(dr.block);
        // If no days exist, set it to a default
        if (meetDays === undefined) {
            meetDays = {
                [SchoolDay.One]: false,
                [SchoolDay.Two]: false,
                [SchoolDay.Three]: false,
                [SchoolDay.Four]: false,
                [SchoolDay.Five]: false,
                [SchoolDay.Six]: false,
                [SchoolDay.Seven]: false,
            };
        }

        // Make sure the dr does not overlap any existing drs for the color
        if (irregularMeetDays({ meets: meetDays }).some(x => irregularMeetDays(dr).includes(x))) {
            problems.addError(dr.uuid, ValidationError.DRConflictWithDR);
        }

        // Save the meeting days by joining them with the others for the same color
        drDays.set(dr.block, irregularMeetJoin(meetDays, dr.meets));

        // Make sure a room has been specified
        if (dr.room.length === 0) {
            problems.addWarn(dr.uuid, ValidationWarning.MissingRoom);
        }

        // Make sure a teacher has been specified
        if (dr.teacher.length === 0) {
            problems.addWarn(dr.uuid, ValidationWarning.MissingTeacher);
        }
    }

    return drDays;
}
