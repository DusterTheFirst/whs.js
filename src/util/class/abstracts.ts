/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { IDR, IElective, IFree, IMajor } from "./blocks";
import { IClassMeta, ITimes } from "./extentions";

/** A type that encompasses any type of class there could be */
export type Class = IElective & IMajor & IDR;
/** A class with all of the information needed to process it */
export type StorageClass = Class & IClassMeta;
export type DisplayClass = (Class | IFree) & ITimes;