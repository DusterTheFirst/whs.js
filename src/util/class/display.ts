/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "../blocks/blockColor";
import { ITimes } from "./extentions";
import { IAdvisedClass } from "./primitives";

/** The display information for a class block */
export interface IClassBlock extends IAdvisedClass, ITimes {
    /** The name of the class block */
    name: string;
    /** The block color which to display for the meets */
    block: BlockColor;
}

/** The most basic information needed to identify a Free block */
export interface IFreeBlock extends ITimes {
    /** The block color which the class meets */
    block: BlockColor;
}