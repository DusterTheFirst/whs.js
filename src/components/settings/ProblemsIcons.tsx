/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import { ValidationError, ValidationErrorMessage, ValidationWarning, ValidationWarningMessage } from "@whsha/classes/v1/class/validation";
import React, { memo } from "react";
import { Alert } from "react-native";
import {ErrorsCount, ProblemsButton, ProblemsIconsView, WarningsCount} from "../../styles/components/problemsicons";
import IconComponent from "../IconComponent";

/** The props for the ProblemsIcons */
interface IProblemsIconsProps {
    /** The errors to show */
    errors?: ValidationError[];
    /** The warnings to show */
    warnings?: ValidationWarning[];
}

/** A component to show the problems and errors with icons */
function ProblemsIcons({ errors = [], warnings = [] }: IProblemsIconsProps) {
    const haswarns = warnings.length > 0;
    const haserrors = errors.length > 0;
    if (!haswarns && !haserrors) {
        return null;
    }

    // TODO: better display
    const showWarns = () => Alert.alert(`${warnings.length} Validation Warning${warnings.length === 1 ? "" : "s"}`, warnings.map(x => `• ${ValidationWarningMessage[x]} (${ValidationWarning[x as unknown as keyof typeof ValidationWarning]})`).join("\n"));
    const showErrors = () => Alert.alert(`${errors.length} Validation Error${errors.length === 1 ? "" : "s"}`, errors.map(x => `• ${ValidationErrorMessage[x]} (${ValidationError[x as unknown as keyof typeof ValidationError]})`).join("\n"));

    return (
        <ProblemsIconsView>
            <ProblemsButton onPress={showWarns} display={haswarns}>
                <WarningsCount>{warnings.length}</WarningsCount>
                <IconComponent name="warning" color="gold" />
            </ProblemsButton>
            <ProblemsButton onPress={showErrors} display={haserrors}>
                <ErrorsCount>{errors.length}</ErrorsCount>
                <IconComponent name="alert" color="red" />
            </ProblemsButton>
        </ProblemsIconsView>
    );
}

export default memo(ProblemsIcons);