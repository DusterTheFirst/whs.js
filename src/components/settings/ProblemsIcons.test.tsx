/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import React from "react";
import { render } from "react-native-testing-library";
import { ValidationError, ValidationWarning } from "../../util/class/validation";
import ProblemsIcons from "./ProblemsIcons";

// TODO: Handle press
describe("Renders different sets of problem icons", () => {
    it("Renders with no problems", () => {
        const comp = render(<ProblemsIcons />);

        expect(comp.toJSON()).toMatchSnapshot();
    });

    it("Renders with only warnings", () => {
        const comp = render(<ProblemsIcons warnings={[ValidationWarning.MissingName, ValidationWarning.MissingRoom]} />);

        expect(comp.toJSON()).toMatchSnapshot();
    });

    it("Renders with only errors", () => {
        const comp = render(<ProblemsIcons errors={[ValidationError.DRConflictWithDR, ValidationError.MeetsEveryDay, ValidationError.MajorHasDuplicateBlockColor]} />);

        expect(comp.toJSON()).toMatchSnapshot();
    });

    it("Renders with both errors and warnings", () => {
        const comp = render(
            <ProblemsIcons
                warnings={[ValidationWarning.MissingName, ValidationWarning.MissingRoom]}
                errors={[ValidationError.DRConflictWithDR, ValidationError.MeetsEveryDay, ValidationError.MajorHasDuplicateBlockColor]}
            />
        );

        expect(comp.toJSON()).toMatchSnapshot();
    });
});