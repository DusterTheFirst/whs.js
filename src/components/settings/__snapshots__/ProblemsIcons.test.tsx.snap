// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders different sets of problem icons Renders with both errors and warnings 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "paddingRight": 10,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "opacity": 1,
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "gold",
            "paddingRight": 5,
          },
        ]
      }
    >
      2
    </Text>
    <View
      style={
        Object {
          "marginBottom": -3,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "opacity": 1,
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "red",
            "paddingRight": 5,
          },
        ]
      }
    >
      3
    </Text>
    <View
      style={
        Object {
          "marginBottom": -3,
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;

exports[`Renders different sets of problem icons Renders with no problems 1`] = `null`;

exports[`Renders different sets of problem icons Renders with only errors 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "paddingRight": 10,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "display": "none",
        "flexDirection": "row",
        "opacity": 1,
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "gold",
            "paddingRight": 5,
          },
        ]
      }
    >
      0
    </Text>
    <View
      style={
        Object {
          "marginBottom": -3,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "opacity": 1,
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "red",
            "paddingRight": 5,
          },
        ]
      }
    >
      3
    </Text>
    <View
      style={
        Object {
          "marginBottom": -3,
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;

exports[`Renders different sets of problem icons Renders with only warnings 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "paddingRight": 10,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "opacity": 1,
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "gold",
            "paddingRight": 5,
          },
        ]
      }
    >
      2
    </Text>
    <View
      style={
        Object {
          "marginBottom": -3,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "display": "none",
        "flexDirection": "row",
        "opacity": 1,
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "red",
            "paddingRight": 5,
          },
        ]
      }
    >
      0
    </Text>
    <View
      style={
        Object {
          "marginBottom": -3,
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;
